### ISSUE TITLE
[Linux] JIT.Regression/CLR-x86-JIT/v2.1/b152292

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#20358]
All JitStress=2 Linux/arm64 runs are failing in b152292.

E.g.,
https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm64/job/jitstress/job/arm64_cross_checked_ubuntu16.04_jitstress2_tst/14/consoleFull
```
13:57:37 
      JIT_Regression._CLR_x86_JIT_v2_1_b152292_b152292_b152292_._CLR_x86_JIT_v2_1_b152292_b152292_b152292_sh [FAIL]
13:57:37         /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_jitstress2_tst/bin/tests/Linux.arm64.Checked/JIT/Regression/CLR-x86-JIT/v2.1/b152292/b152292/b152292.sh: line 244: 39135 Aborted                 (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
13:57:37         /n/nReturn code:      1/nRaw output file:      /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_jitstress2_tst/bin/tests/Linux.arm64.Checked/JIT/Regression/Reports/JIT.Regression/CLR-x86-JIT/v2.1/b152292/b152292/b152292.output.txt/nRaw output:/nBEGIN EXECUTION
13:57:37         /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_jitstress2_tst/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun b152292.exe
13:57:37         Expected: 100
13:57:37         Actual: 134
13:57:37         END EXECUTION - FAILED
13:57:37         Test Harness Exitcode is : 1
13:57:37         /nTo run the test:/n> set CORE_ROOT=/home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_jitstress2_tst/bin/tests/Linux.arm64.Checked/Tests/Core_Root/n> /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_jitstress2_tst/bin/tests/Linux.arm64.Checked/JIT/Regression/CLR-x86-JIT/v2.1/b152292/b152292/b152292.sh/n
13:57:37         Expected: True
13:57:37         Actual:   False
13:57:38         Stack Trace:
13:57:38           /mnt/j/workspace/dotnet_coreclr/master/arm64_cross_checked_ubuntu16.04/bin/tests/Linux.arm64.Checked/TestWrappers/JIT.Regression/JIT.Regression.XUnitWrapper.cs(61639,0): at JIT_Regression._CLR_x86_JIT_v2_1_b152292_b152292_b152292_._CLR_x86_JIT_v2_1_b152292_b152292_b152292_sh()
```

**Comment by BruceForstall** [comment]
They are also failing for arm32: https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/arm_cross_checked_ubuntu_jitstress2_tst/35/consoleFull

**Comment by echesakovMSFT** [comment]
Also fails on Linux/x64 with the following environment:
```bash
export COMPlus_TieredCompilation=0
export COMPlus_JitStress=2
```
https://mc.dot.net/#/user/coreclr-pulls/pr~2Fdotnet~2Fcoreclr~2Frefs~2Fheads~2Fmaster/test~2Ffunctional~2Fcli~2Fpri1/20190203.728/workItem/JIT.Regression/analysis/xunit/JIT_Regression._CLR_x86_JIT_v2_1_b152292_b152292_b152292_~2F_CLR_x86_JIT_v2_1_b152292_b152292_b152292_sh

```
+ /home/helixbot/dotnetbuild/work/f211b02a-9804-4db1-a7ed-a101bb18c38f/Payload/corerun /home/helixbot/dotnetbuild/work/f211b02a-9804-4db1-a7ed-a101bb18c38f/Payload/xunit.console.dll JIT.Regression.XUnitWrapper.dll -parallel collections -nocolor -noshadow -xml testResults.xml -notrait category=outerloop -notrait category=failing
xUnit.net Console Runner v2.4.1 (64-bit .NET Core 4.6.27403.0)
  Discovering: JIT.Regression.XUnitWrapper
  Discovered:  JIT.Regression.XUnitWrapper
  Starting:    JIT.Regression.XUnitWrapper
    JIT_Regression._CLR_x86_JIT_v2_1_b152292_b152292_b152292_._CLR_x86_JIT_v2_1_b152292_b152292_b152292_sh [FAIL]
      /home/helixbot/dotnetbuild/work/f211b02a-9804-4db1-a7ed-a101bb18c38f/Work/7e8afef3-afb9-4b02-b9db-de8c07be84a8/Exec/CLR-x86-JIT/v2.1/b152292/b152292/b152292.sh: line 244:  4081 Aborted                 (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
      /n/nReturn code:      1/nRaw output file:      /home/helixbot/dotnetbuild/work/f211b02a-9804-4db1-a7ed-a101bb18c38f/Work/7e8afef3-afb9-4b02-b9db-de8c07be84a8/Exec/Reports/JIT.Regression/CLR-x86-JIT/v2.1/b152292/b152292/b152292.output.txt/nRaw output:/nBEGIN EXECUTION
      /home/helixbot/dotnetbuild/work/f211b02a-9804-4db1-a7ed-a101bb18c38f/Payload/corerun b152292.exe
      Expected: 100
      Actual: 134
      END EXECUTION - FAILED
      Test Harness Exitcode is : 1
      /nTo run the test:/n> set CORE_ROOT=/home/helixbot/dotnetbuild/work/f211b02a-9804-4db1-a7ed-a101bb18c38f/Payload/n> /home/helixbot/dotnetbuild/work/f211b02a-9804-4db1-a7ed-a101bb18c38f/Work/7e8afef3-afb9-4b02-b9db-de8c07be84a8/Exec/CLR-x86-JIT/v2.1/b152292/b152292/b152292.sh/n
      Expected: True
      Actual:   False
      Stack Trace:
        /__w/1/s/bin/tests/Linux.x64.Checked/TestWrappers/JIT.Regression/JIT.Regression.XUnitWrapper.cs(61569,0): at JIT_Regression._CLR_x86_JIT_v2_1_b152292_b152292_b152292_._CLR_x86_JIT_v2_1_b152292_b152292_b152292_sh()
  Finished:    JIT.Regression.XUnitWrapper
=== TEST EXECUTION SUMMARY ===
   JIT.Regression.XUnitWrapper  Total: 1360, Errors: 0, Failed: 1, Skipped: 0, Time: 390.144s
```

**Comment by erozenfeld** [comment]
This is a bug in optcse and/or value numbering. It loses a nullcheck when replacing use with a def. Here is the relevant part of the jit dump:

```
************ Trees at start of optValnumCSE_Heuristic()

------------ BB01 [000..007) (return), preds={} succs={}

***** BB01, stmt 1
     (  7, 10)              [000004] ------------              *  STMT      void  (IL 0x000...0x006)
N008 (  7, 10)              [000003] ---XG-------              \--*  RETURN    int    $201
N006 (  4,  7) CSE #01 (def)[000011] ---X--------                 |  /--*  IND       int    <l:$1c3, c:$1c2>
N004 (  1,  4)              [000008] ------------                 |  |  |  /--*  CNS_INT   long   0x109A4 field offset Fseq[i16999] $100
N005 (  2,  5)              [000009] -------N----                 |  |  \--*  ADD       byref  $140
N003 (  1,  1)              [000007] ------------                 |  |     \--*  LCL_VAR   ref    V00 arg0         u:1 (last use) $80
N007 (  6,  9) CSE #01 (use)[000010] ---XG-------                 \--*  COMMA     int    <l:$1c3, c:$1c2>
N002 (  2,  2)              [000006] ---X---N----                    \--*  NULLCHECK byte   $c1
N001 (  1,  1)              [000005] ------------                       \--*  LCL_VAR   ref    V00 arg0         u:1 $80

-------------------------------------------------------------------------------------------------------------------


Aggressive CSE Promotion cutoff is 100
Moderate CSE Promotion cutoff is 50
Framesize estimate is 0x0000
We have a small frame

Sorted CSE candidates:
CSE #01, {$1c1, $c0 } cseMask=0000000000000001,useCnt=1: [def=100, use=100] :: N006 (  4,  7) CSE #01 (def)[000011] ---X--------              *  IND       int    <l:$1c3, c:$1c2>


Considering CSE #01 {$1c1, $c0 } [def=100, use=100, cost= 4] CSE Expression:
N006 (  4,  7) CSE #01 (def)[000011] ---X--------              *  IND       int    <l:$1c3, c:$1c2>
N004 (  1,  4)              [000008] ------------              |  /--*  CNS_INT   long   0x109A4 field offset Fseq[i16999] $100
N005 (  2,  5)              [000009] -------N----              \--*  ADD       byref  $140
N003 (  1,  1)              [000007] ------------                 \--*  LCL_VAR   ref    V00 arg0         u:1 (last use) $80



*** JitStress: STRESS_MAKE_CSE ***

JitStressBiasedCSE is OFF, but JitStress is ON: generated bias=61.
Promoting CSE because gen=6 < bias=61

Promoting CSE:

lvaGrabTemp returning 2 (V02 rat0) (a long lifetime temp) called for ValNumCSE.

CSE #01 def at [000011] replaced in BB01 with def of V02
optValnumCSE morphed tree:
N012 ( 14, 15)              [000003] -A-XG-------              *  RETURN    int    $201
N009 (  3,  2)              [000014] ------------              |     /--*  LCL_VAR   int    V02 cse0          <l:$1c3, c:$1c2>
N010 ( 11, 12)              [000015] -A-X--------              |  /--*  COMMA     int    <l:$1c3, c:$1c2>
N006 (  4,  7)              [000011] ---X--------              |  |  |  /--*  IND       int    <l:$1c3, c:$1c2>
N004 (  1,  4)              [000008] ------------              |  |  |  |  |  /--*  CNS_INT   long   0x109A4 field offset Fseq[i16999] $100
N005 (  2,  5)              [000009] -------N----              |  |  |  |  \--*  ADD       byref  $140
N003 (  1,  1)              [000007] ------------              |  |  |  |     \--*  LCL_VAR   ref    V00 arg0         u:1 (last use) $80
N008 (  8, 10)              [000013] -A-X----R---              |  |  \--*  ASG       int    $VN.Void
N007 (  3,  2)              [000012] D------N----              |  |     \--*  LCL_VAR   int    V02 cse0          <l:$1c3, c:$1c2>
N011 ( 13, 14) CSE #01 (use)[000010] -A-XG-------              \--*  COMMA     int    <l:$1c3, c:$1c2>
N002 (  2,  2)              [000006] ---X---N----                 \--*  NULLCHECK byte   $c1
N001 (  1,  1)              [000005] ------------                    \--*  LCL_VAR   ref    V00 arg0         u:1 $80


Working on the replacement of the CSE #01 use at [000010] in BB01

This CSE use has side effects and/or nested CSE defs. The sideEffectList:
N006 (  4,  7)              [000011] ---X--------              /--*  IND       int    <l:$1c3, c:$1c2>
N004 (  1,  4)              [000008] ------------              |  |  /--*  CNS_INT   long   0x109A4 field offset Fseq[i16999] $100
N005 (  2,  5)              [000009] -------N----              |  \--*  ADD       byref  $140
N003 (  1,  1)              [000007] ------------              |     \--*  LCL_VAR   ref    V00 arg0         u:1 (last use) $80
N008 (  8, 10)              [000013] -A-X----R---              *  ASG       int    $VN.Void
N007 (  3,  2)              [000012] D------N----              \--*  LCL_VAR   int    V02 cse0          <l:$1c3, c:$1c2>

optValnumCSE morphed tree:
N009 ( 12, 13)              [000003] -A-XG-------              *  RETURN    int    $201
N007 (  3,  2)              [000016] ------------              |  /--*  LCL_VAR   int    V02 cse0          <l:$1c1, c:$200>
N008 ( 11, 12)              [000017] -A-X--------              \--*  COMMA     int    <l:$1c3, c:$1c2>
N004 (  4,  7)              [000011] ---X--------                 |  /--*  IND       int    <l:$1c3, c:$1c2>
N002 (  1,  4)              [000008] ------------                 |  |  |  /--*  CNS_INT   long   0x109A4 field offset Fseq[i16999] $100
N003 (  2,  5)              [000009] -------N----                 |  |  \--*  ADD       byref  $140
N001 (  1,  1)              [000007] ------------                 |  |     \--*  LCL_VAR   ref    V00 arg0         u:1 (last use) $80
N006 (  8, 10)              [000013] -A-X----R---                 \--*  ASG       int    $VN.Void
N005 (  3,  2)              [000012] D------N----                    \--*  LCL_VAR   int    V02 cse0          <l:$1c3, c:$1c2>

```

Note that 

```
N008 ( 11, 12)              [000017] -A-X--------              \--*  COMMA     int    <l:$1c3, c:$1c2>
```
and
```
N006 (  4,  7)              [000011] ---X--------              /--*  IND       int    <l:$1c3, c:$1c2>
```

have the same value numbers even though it's possible that they have different exception behavior.

Then the call to extract side effects only extracts persistent side effects so exception-causing side effects are not extracted:
https://github.com/dotnet/coreclr/blob/6baf0f9d160ad1802f5d8c98d55c44a4550f6119/src/jit/optcse.cpp#L2204

and
```
N002 (  2,  2)              [000006] ---X---N----                 \--*  NULLCHECK byte   $c1
N001 (  1,  1)              [000005] ------------                    \--*  LCL_VAR   ref    V00 arg0         u:1 $80
```
is gone after the CSE replacement.

**Comment by erozenfeld** [comment]
Some context: the test just tests that NullReferenceException is thrown when trying to access a filed at a large offset. Explicit NULLCHECK is inserted and shouldn't be removed.

**Comment by erozenfeld** [comment]
One possible fix for this test is to mark the COMMA (and not just NULLCHECK) with GTF_DONT_CSE but I think that's just a workaround.

Currently we assume that if two expressions produce the same value and may throw the same set of exceptions we can CSE them. But I don't think we don't model that for some value of an operand one expression may throw an exception while the other won't (or throw a different exception).

**Comment by mikedn** [comment]
> But I don't think we model that for some value of an operand one expression may throw an exception while the other won't (or throw a different exception).

@briansull VN exceptions works should have taken care of that. COMMA value numbering does preserve exceptions sets from both `op1`: https://github.com/dotnet/coreclr/blob/be1b5db77a593fdea460ef7ab90e63b3817e7853/src/jit/valuenum.cpp#L8206-L8228

**Comment by erozenfeld** [comment]
I don't think the current model with exception sets is sound.

In the bug's example we have
expr1: 
COMMA(NULLCHECK(x), INDIR(x+<large_offset>))
expr2: 
INDIR(x+<large_offset>)

They both produce the same value and have the same set of exceptions {NullReferenceException}. Yet, it's not correct to CSE them.

And I think this model may break in other cases as well. What if we have
expr1:
COMMA(DIV(1/x), COMMA(INDIR(y), z)))
expr2:
COMMA(INDIR(y), COMMA(DIV(1/x), z)))

Here again the values are the same and the exception sets are the same: {DivideByZeroException, NullReferenceException}. Yet with x=0 and y=0 expr1 will throw DivideByZeroException and expr2 will throw NullReferenceException so we shouldn't be able to CSE expr1 and expr2.

@briansull Am I missing something?

**Comment by erozenfeld** [comment]
Even simpler example:
expr1:
COMMA(INDIR(x), z)
expr2:
COMMA(INDIR(y), z)

Again, same value, same exception sets, but we obviously shouldn't be able to CSE them.

**Comment by erozenfeld** [comment]
@dotnet/jit-contrib

**Comment by erozenfeld** [comment]
The bug causing this issue is in `fgValueNumberAddExceptionSetForIndirection`:
these lines 
https://github.com/dotnet/coreclr/blob/de287410d790bc93668d4d01a202494087af8885/src/jit/valuenum.cpp#L9242-L9249
need to be moved to the end of the loop.

The same for these lines in the second loop:

https://github.com/dotnet/coreclr/blob/de287410d790bc93668d4d01a202494087af8885/src/jit/valuenum.cpp#L9272-L9279

**Comment by erozenfeld** [comment]
I'll have a PR out shortly.

**Comment by briansull** [comment]
@erozenfeld It looks like you have correctly identified the problem (and the fix)

**Comment by erozenfeld** [comment]
@briansull What do you think about this example from my comments above? Is this another bug?
>> expr1:
COMMA(DIV(1/x), COMMA(INDIR(y), z)))
expr2:
COMMA(INDIR(y), COMMA(DIV(1/x), z)))
Here again the values are the same and the exception sets are the same: {DivideByZeroException, NullReferenceException}. Yet with x=0 and y=0 expr1 will throw DivideByZeroException and expr2 will throw NullReferenceException so we shouldn't be able to CSE expr1 and expr2.

Do we need ordered sets to model this correctly?

**Comment by briansull** [comment]
I wouldn't want to change exception sets to be ordered.  I will have to think about this issue.

**Comment by briansull** [comment]
I don't believe that the ordering of child nodes here in a GT_COMMA or those in other nodes such a GT_ADD are a problem.  In order to use this VN in a CSE def both child operand have to be fully evaluated.
So if either side throws an exception the it will happen at the CSE def.

**Comment by erozenfeld** [comment]
I started trying to create a repro for a possible exception ordering problem because of CSE and discovered that we have a problem even without CSE/value numbering involved. In this simple example

```
[MethodImpl(MethodImplOptions.NoInlining)]
    static int foo(Test t, int i)
    {
        int x = t.f + 1 / i;
        return x;
    }
```
when passed null, 0 we throw `DivideByZeroException` instead of `NullReferenceException` (operands of addition have to be evaluated left to right). I'll open a separate issue for that.

I'm still not convinced our value numbering model is sound wrt to exception ordering but I want to fix this issue first. Then I'll try to construct an example involving value numbering and CSE.

**Comment by mikedn** [comment]
In that example it's lowering's containment code that somehow manages to move the indir past div. AFAIR that happened before, though I thought it was fixed.

**Comment by erozenfeld** [comment]
#22556

**Comment by erozenfeld** [comment]
Yes, the bug is in lowering containment code.

**Comment by erozenfeld** [comment]
I thought some more about exception ordering and I think our CSE is ok. We would have a problem if we tried to do code hoisting optimization (in Muchnick's terms), which finds expressions that are always evaluated following some point in a program, regardless of the execution path, and moves them to the latest point beyond which they would always be evaluated. In this case exception ordering would matter.

**Comment by mikedn** [comment]
> We would have a problem if we tried to do code hoisting optimization (in Muchnick's terms), which finds expressions that are always evaluated following some point in a program, regardless of the execution path, and moves them to the latest point beyond which they would always be evaluated.

The JIT doesn't do that but beware that it does loop hoisting and it gets the order wrong: #7147

**Comment by erozenfeld** [comment]
Thank you for reminding me about that bug. I hope I'll get to it soon. Not sure why we haven't fixed it in 2.5 years and why it's still marked Future.

**Comment by mikedn** [comment]
> I hope I'll get to it soon

Working on it, I'm mostly done with the `swap` case: #20650
